added url conversion

kevinpapst opened the pull request:
Fixes https://github.com/FreshRSS/FreshRSS/issues/2172

kevinpapst committed:
added url conversion

meminens commented:
Am I supposed to edit any core FreshRSS files for this PR to work? I am using the updated extension.php file in the YouTube extension folder but I am getting the following error when I attempt to add RSS using the standard YouTube channel URL format (e.g. https://www.youtube.com/channel/UCp_maMdOnRVq3Ca45ILwsOg).

```
2018-12-08 20:44:53A feed could not be found at `https://www.youtube.com/channel/UCp_maMdOnRVq3Ca45ILwsOg`; the status code is `200` and content-type is `text/html; charset=utf-8` [https://www.youtube.com/channel/UCp_maMdOnRVq3Ca45ILwsOg]
```

Confirmed that the YouTube extension is enabled and functioning normally.

kevinpapst commented:
No, it should work out-of-the-box. But for some reason a couple of channel URLs do not work, like your example.
I grabbed some channel and user URLs from the Youtube startpage and they work, for example:
https://www.youtube.com/user/trickstutorials
https://www.youtube.com/channel/UCe8Rw9XQcT3zyXoHiMqqdzw

meminens commented:
Thanks! Yes, I have also been able to add those example YT channels. Interesting that some channel URLs behave differently.

marienfressinaud commented:
Hi @kevinpapst! Do you remember what prevent you to merge this PR? I'd like to install it for my FRSS instance and this feature would be very valuable. There's an alternative in https://github.com/cn-tools/cntools_FreshRssExtensions but I'd like to avoid multiple Youtube extensions and your solution looks more robust :)

kevinpapst commented:
I guess I wanted to test some more channels and then I forgot about the PR^^

Is the branch working for you? If so, I don't mind to merge it.

marienfressinaud commented on xExtension-YouTube/extension.php:
```diff
@@ -36,9 +36,29 @@ class YouTubeExtension extends Minz_Extension
+        $matches = [];
+
+        if (preg_match('#^https?://www\.youtube\.com/channel/([0-9a-zA-Z_-]{6,36})$#', $feed->url(), $matches) !== 1) {
+            if (preg_match('#^https?://www\.youtube\.com/user/([0-9a-zA-Z_-0]{6,36})$#', $feed->url(), $matches) !== 1) {
```
The last "0" in `[0-9a-zA-Z_-0]` should be removed

marienfressinaud commented on xExtension-YouTube/extension.php:
```diff
@@ -36,9 +36,29 @@ class YouTubeExtension extends Minz_Extension
+            }
+        }
+
+        $feed->_url('https://www.youtube.com/feeds/videos.xml?channel_id=' . $matches[1]);
```
For `/user/` urls, I think the correct feed URL is `https://www.youtube.com/feeds/videos.xml?user=xxx`. It works with your example (in comments) because I think SimplePie got redirected by Youtube to a `channel_id` type URL

marienfressinaud reviewed (commented):
Some channels are not working because Youtube redirects SimplePie to a different page. I don't know why, maybe it's to prevent robots to grab content.

I think you can use the `check_url_before_add` hook instead: it's a function that take the feed URL and return a URL

kevinpapst commented:
This might be too much to ask, but I don't have a FRSS installation available currently and can't test it. Are you able/Would you mind to fork it, make the changes and sent them in?

marienfressinaud commented:
Yep, no problem! :)

marienfressinaud committed:
Convert URL on check_url_before_add (#12)

SimplePie is redirected by Youtube on some channels, and we might end up
on a totally different page. By using the `check_url_before_add` hook,
we change the URL before the call to SimplePie.

kevinpapst commented:
Thanks @marienfressinaud for testing and the update!

marienfressinaud commented:
You're welcome! I'm opening another PR to fix some unrelated details

kevinpapst commented:
Soon? If so, I will wait with the 0.8 release.

kevinpapst commented:
Forget it, hitting the release button once more doesn't cost any money 😁
